src/perception/aruco_processing.cpp

```diff
@@ -0,0 +1,136 @@
+#include "aruco_detect.hpp"
+
+#include "filter.hpp"
+
+// TODO: add cache logic to filter out false positives
+
+/**
+ * Detect fiducials from raw image using OpenCV and calculate their screen space centers.
+ * Maintain the immediate buffer - holds all fiducials seen currently on screen.
+ * Later camera space pose information is filled in when we receive point cloud data.
+ *
+ * @param msg
+ */
+void FiducialsNode::imageCallback(sensor_msgs::ImageConstPtr const& msg) {
+    if (!mEnableDetections) return;
+
+    if (mIsVerbose) {
+        ROS_INFO("Got image %d", msg->header.seq);
+    }
+
+    try {
+        mCvPtr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
+
+        // Detect the fiducial vertices in screen space and their respective ids
+        cv::aruco::detectMarkers(mCvPtr->image, mDictionary, mCornersCache, mIdsCache, mDetectorParams);
+
+        // Save fiducials currently on screen to the immediate buffer
+        for (size_t i = 0; i < mIdsCache.size(); ++i) {
+            int id = mIdsCache[i];
+            cv::Point2f imageCenter = std::accumulate(mCornersCache[i].begin(), mCornersCache[i].end(), cv::Point2f{}) / 4.0f;
+            mImmediateFiducials[id].id = id;
+            mImmediateFiducials[id].imageCenter = imageCenter;
+        }
+
+        // Remove fiducials in the immediate buffer that are no longer in sight
+        auto it = mImmediateFiducials.begin();
+        while (it != mImmediateFiducials.end()) {
```

Comment:
yeah in this case I just preferred readable logic

Reply:
In that case, why did you add new detections before removing stale ones? I think I agree with Ashwin that it may be clearer to remove before adding, unless I'm missing what both of you are talking about?

Reply:
Sure we can do it that way

Reply:
Also you realize this is `std::unordered_map` right, not `std::vector`

Reply:
Done